drivers: timer: extend nrf_rtc_timer to handle overflow

pkwiek opened the pull request:
nrf_rtc_timer is extended with a capability to handle RTC overflow,
allowing it to operate on absolute RTC ticks, rather than relative
ticks. The overflow handling is ZLI-proof, due to use of exclusive
access instructions instead of Zephyr's synchronization primitives.
Callbacks are given an absolute RTC tick value as a parameter instead
of CC register's value.

Signed-off-by: Jedrzej Ciupis <jedrzej.ciupis@nordicsemi.no>
Signed-off-by: Paweł Kwiek <pawel.kwiek@nordicsemi.no>

Commit:
drivers: timer: extend nrf_rtc_timer

This commit introduces the following changes:

* nrf_rtc_timer is extended with a capability to handle RTC overflow,
allowing it to operate on absolute RTC ticks, rather than relative
ticks.
* overflow handling is ZLI-proof and relies on the sys clock
handler being executed twice every RTC counter's overflow.
* callbacks are given an absolute RTC tick value as a parameter instead
of CC register's value. The absolute RTC tick value is the RTC counter
value set during CC channel configuration extended to 64 bits.
* in case the timer's target time is in the past or is the current tick,
the timer fires as soon as possible, however still from the RTC's ISR
context.
* in case an active timer is set again with the same target time, it is
not scheduled again - only its event data is updated. Otherwise, the
timer is scheduled as usual.
* a scheduled timer can be aborted.
* system clock functions are now using 64 bit values internally.

Signed-off-by: Andrzej Kuroś <andrzej.kuros@nordicsemi.no>
Signed-off-by: Jedrzej Ciupis <jedrzej.ciupis@nordicsemi.no>
Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>
Signed-off-by: Paweł Kwiek <pawel.kwiek@nordicsemi.no>

ankuns commented on tests/drivers/timer/nrf_rtc_timer/src/main.c:
```diff
@@ -53,34 +54,48 @@ static void stop_zli_timer0(void)
-		"Handler called in wrong time (%d), set cc: %d, got cc: %d",
-		now, data->cc_val, cc_value);
+		"Handler called in wrong time (%d), set time: %d, got time: %d",
+		now, data->target_time, target_time);
```
`%d` format string does not match `uint64_t`

ankuns commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -238,12 +431,56 @@ static void sys_clock_timeout_handler(int32_t chan,
+		overflow_and_counter_get(&overflow, &counter);
+
+		curr_time =
+		    overflow_and_counter_to_target_time(overflow, counter);
```
You could defer this calculation until `handler` is called.
line 480:
```c
if (shall_execute_handler) {
    uint64_t t = overflow_and_counter_to_target_time(overflow, counter);
    handler(chan, t, user_context);
}

ankuns commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -238,12 +431,56 @@ static void sys_clock_timeout_handler(int32_t chan,
+			user_context = cc_data[chan].user_context;
+			cc_data[chan].callback = NULL;
+
+			shall_execute_handler = handler != NULL;
```
You can give up with separate variable `shall_execute_handler`.
Let's simply have initial value in line 447
```c
z_nrf_rtc_timer_compare_handler_t handler = NULL
```
then
```c
if (handler)
{
    // ... 
    handler()
}
```

ankuns commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -238,12 +431,56 @@ static void sys_clock_timeout_handler(int32_t chan,
+		/* If target_cc is in the past or is equal to current counter
+		 * value, execute the handler.
+		 */
+		if (counter_sub(cc_data[chan].target_cc, counter + 1)
```
Why is `+1` in `counter + 1` needed? can't following do the job?
```c
if (counter_sub(counter, cc_data[chan].target_cc) < COUNTER_HALF_SPAN)
```
This includes the case when counter == target_cc

ankuns commented on include/drivers/timer/nrf_rtc_timer.h:
```diff
@@ -11,8 +11,12 @@
+
 typedef void (*z_nrf_rtc_timer_compare_handler_t)(int32_t id,
-						uint32_t cc_value,
+						uint64_t cc_value,
```
Let's decide if we are going to provide `cc_value` or `current_time` at which we real triggered.
We are going to provide `current_time` to not have to read timer again in callback.
Let's name this parameter `current_time`, `curr_time` or `now_time` but not `cc_value`.

ankuns reviewed (changes requested):
First round

nordic-krch reviewed (commented):
few comments.

e-rk reviewed (changes requested):
First round of review.

pkwiek committed:
modules: hal_nordic: align & extend lp_timer platform

This commit aligns nrf_802154_lp_timer_zephyr platform to new API of
nrf_rtc_timer. Moreover, it implements lp_timer synchronization
functions using the new nrf_rtc_timer API.

Signed-off-by: Andrzej Kuroś <andrzej.kuros@nordicsemi.no>
Signed-off-by: Jedrzej Ciupis <jedrzej.ciupis@nordicsemi.no>
Signed-off-by: Paweł Kwiek <pawel.kwiek@nordicsemi.no>

nordic-krch reviewed (commented):
Couple of comments.

nordic-krch reviewed (changes requested):
Accidentally pressed approval.

e-rk reviewed (approved):
Didn't review the latest changes, but I don't want github to block this PR any longer.

nordic-krch commented:
Please update commit messages. There is a typo in `modules: hal_nordic: Align lptimer platform do nrf_rtc_timer ` and second commit mentions extending timeouts beyond half span.

nordic-krch reviewed (approved):
Looks good. One small thing: commit message `modules: hal_nordic: Align lptimer platform do nrf_rtc_timer` should probably be `modules: hal_nordic: Align lptimer platform to nrf_rtc_timer`

carlescufi commented:
@pkwiek please take a look at the buildkite issues:

```
/workdir/zephyr/drivers/timer/nrf_rtc_timer.c:80:13: error: 'event_disable' defined but not used [-Werror=unused-function]
--
  | 80 \| static void event_disable(int32_t chan)
  | \|             ^~~~~~~~~~~~~
```

nordic-krch commented on include/drivers/timer/nrf_rtc_timer.h:
```diff
@@ -11,8 +11,25 @@
 
+/** @brief Maximum allowed time span that is considered to be in the future.
+ */
+#define NRF_RTC_TIMER_MAX_SCHEDULE_SPAN (BIT(24) - 1)
```
this is no longer true as mechanism for scheduling alarms longer than half span got removed.

jciupis replied:
Fixed.
